controller/engine_controller.go

```diff
@@ -527,19 +527,32 @@ func (ec *EngineController) DeleteInstance(obj interface{}) error {
 
 	logrus.Infof("Deleting engine instance %v", e.Name)
 
+	defer func() {
+		logrus.WithError(err).Warnf("Failed to delete engine %v for volume %v", e.Name, v.Name)
+		if isRWXVolume && im.Status.CurrentState != longhorn.InstanceManagerStateRunning {
+			// Try the best to delete engine instance.
+			// To prevent that the volume is stuck at detaching state, ignore the error when volume is
+			// a RWX volume and the instance manager is not running.
+			logrus.Warnf("Ignore the failure of deleting engine %v for volume %v", e.Name, v.Name)
+			err = nil
+		}
+	}()
+
 	// For the engine process in instance manager v0.7.0, we need to use the cmdline to delete the process
 	// and stop the iscsi
 	if im.Status.APIVersion == engineapi.IncompatibleInstanceManagerAPIVersion {
 		url := imutil.GetURL(im.Status.IP, engineapi.InstanceManagerDefaultPort)
 		args := []string{"--url", url, "engine", "delete", "--name", e.Name}
 
-		if _, err := util.ExecuteWithoutTimeout([]string{}, engineapi.GetDeprecatedInstanceManagerBinary(e.Status.CurrentImage), args...); err != nil && !types.ErrorIsNotFound(err) {
+		_, err = util.ExecuteWithoutTimeout([]string{}, engineapi.GetDeprecatedInstanceManagerBinary(e.Status.CurrentImage), args...)
+		if err != nil && !types.ErrorIsNotFound(err) {
```

Comment:
What you mentioned is the original implementation.
The purpose of the deletion is try to delete the engine instance and make the failover faster.

Reply:
I mean. 

Because the node is really down, the IM status should become `InstanceManagerStateError` eventually due to the pod was disappeared. Then, if the volume is RWX, we should be able to understand whether the engine was unable to reach is caused by the Pod disappearing.

Of course, we can't guarantee how much time the reconciling will take effect, so there would be some latency to let the volume failover. This is a disadvantage.

@derekbit right? 

cc @shuo-wu

Reply:
After discussing with @derekbit , I think best-effort deletion is fine, but there will be two cases we need to clarify. 
- if somehow the pod is around but is unable to reach to cause the deletion failure, there will be an orphaned engine process after the volume detached, so how the orphaned engine will be tackled? 
- if the node is really down and the pod is gone, then best-effort deletion works perfectly.

For item 1, we just need to have a comment to explain a bit or even if this is an intended leak.

Reply:
Update the comment explaining what will happen if the engine instance is not deleted.

Reply:
Sorry... Will we blindly try to delete the instance?
https://github.com/longhorn/longhorn-manager/pull/1757#discussion_r1138026034